### PR TITLE
[python cli] 'rtw docker enter'

### PR DESCRIPTION
**Pull request by muritane**
just run this `cd ros_team_workspace/rtwcli/ && pip3 install -r requirements.txt` and start new terminal as sourcing is already done in _ros_team_workspace/setup.bash_

@destogl pep257 is commented out temporarily as the import of Set raises Exception: `cannot import name 'Set' from 'collections'`, is there a repo that supports newer python versions?

### CODE REVIEW

**Comment by mamueluth** [comment]
Nice work! Looks good:) 
I would however advise to first create only the cli and call the existing bash scripts with the expected arguments. I would not start recreating/adding logic in the python cli. This should in my opinion be done in a separate step while writing tests at the same time.
Moreover, in my opinion, we have a lot of common functionality which can be extracted from the scripts and collected at one point. I would keep the cli as thin as possible and not scatter to much logic all over the place (otherwise we end up with lot of duplicate logic.)

**Comment by mamueluth** [comment]
Have to give it a closer review

**Comment by destogl** [comment]
> @destogl pep257 is commented out temporarily as the import of Set raises Exception: cannot import name 'Set' from 'collections', is there a repo that supports newer python versions?

I fixed that.

File: rtwcli/.gitignore
Comment (destogl) [inline rtwcli/.gitignore:1]: Please use `.gitignore` inside your $HOME folder. We can also add this into templates for setting up default fiel when setting up RTW.

File: rtwcli/rtwcli/completion/rtw-argcomplete.bash [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/completion/rtw-argcomplete.bash:1]: Add our license.

File: rtwcli/rtwcli/rtwcli/cli.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# Copyright 2016-2017 Dirk Thomas
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/cli.py:2]: Add our license

File: rtwcli/rtwcli/rtwcli/cli.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import argparse
+import signal
+
+from rtwcli.command import add_subparsers_on_demand
+
+
+def main(*, script_name="rtw", argv=None, description=None, extension=None):
+    if description is None:
+        description = f"{script_name} is an extensible command-line tool " "for ROS Team WS."
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/cli.py:24]:
```suggestion
        description = f"{script_name} is an extensible command-line tool " "for ROS Team Workspace."
```

File: rtwcli/rtwcli/rtwcli/command/__init__.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
+# Copyright 2016-2017 Dirk Thomas
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/command/__init__.py:2]: our license

File: rtwcli/rtwcli/rtwcli/command/__init__.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    :param dict command_extensions: dict of command extensions by their name
+      where each contributes a command with specific arguments
+    """
+    import warnings
+
+    warnings.warn(
+        "'rtwcli.command.add_subparsers' is deprecated, use "
+        "'rtwcli.command.add_subparsers_on_demand' instead",
+        stacklevel=2,
+    )
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/command/__init__.py:96]: Do we need this depreacation notice?

File: rtwcli/rtwcli/rtwcli/command/extension_points.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2016-2017 Dirk Thomas
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/command/extension_points.py:2]: our license

File: rtwcli/rtwcli/rtwcli/command/extensions.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
+# Copyright 2016-2017 Dirk Thomas
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/command/extensions.py:2]: our license

File: rtwcli/rtwcli/rtwcli/entry_points.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+# Copyright 2016-2017 Dirk Thomas
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/entry_points.py:2]: our license

File: rtwcli/rtwcli/rtwcli/plugin_system.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
+# Copyright 2016-2017 Dirk Thomas
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwcli/rtwcli/plugin_system.py:2]: our license

File: rtwcli/rtwdocker/rtwdocker/__init__.py
Comment (destogl) [inline rtwcli/rtwdocker/rtwdocker/__init__.py:1]: can we use `rtw_docker` as a package name? I find it easier to read and faster to see what is happening.

File: rtwcli/rtwdocker/rtwdocker/api/__init__.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def start_container(container_instance_name: str):
+    if not container_instance_name:
+        print("No container name given. Can not start docker instance.")
+        return
+
+    print(f"Starting container instance: {container_instance_name}")
+    subprocess.run(["xhost", "+local:docker"])
+    subprocess.run(["docker", "start", container_instance_name])
```
Comment (destogl) [inline rtwcli/rtwdocker/rtwdocker/api/__init__.py:26]: we should also fallback here to use existing rtw command. Then we move one by one - removing the old one and adding new one

File: rtwcli/rtwdocker/rtwdocker/verb/__init__.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwdocker/rtwdocker/verb/__init__.py:1]: license

File: rtwcli/rtwpkg/rtwpkg/verb/__init__.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (destogl) [inline rtwcli/rtwpkg/rtwpkg/verb/__init__.py:1]: license

**Review by destogl — CHANGES_REQUESTED** [review]
As much as I agree with @mamueluth, we should keep effort in reusing bash commands low.

In general, this looks excellent to me. Still, there are a few open questions to clarify:

1. Can we reduce the number of folders and files by adding all commands into one package? (so we end up having only two packages?)
2. What will happen if I execute `auto-source` one more time? will I break my current setup?
3. we should split this into multiple PRs. - 1. with base setup also initial command, 2. with dokcer setup, 3. with package setup.
4. Initial setup should add `info` command giving info output about rtw setup. For now let just output the base path of the package. We can add later other stuff.

File: rtwcli/rtwcli/completion/rtw-argcomplete.bash (hunk 1 above)
Comment (destogl) [reply rtwcli/rtwcli/completion/rtw-argcomplete.bash:1]:
```suggestion
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright (c) 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwcli/completion/rtw-argcomplete.zsh [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
[the remark this answers is not in the document]
Comment (destogl) [reply rtwcli/rtwcli/completion/rtw-argcomplete.zsh:1]:
```suggestion
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright (c) 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwcli/rtwcli/cli.py (hunk 2 above)
Comment (destogl) [reply rtwcli/rtwcli/rtwcli/cli.py:2]:
```suggestion
# Copyright 2016-2017 Dirk Thomas
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright (c) 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwcli/rtwcli/plugin_system.py (hunk 9 above)
Comment (destogl) [reply rtwcli/rtwcli/rtwcli/plugin_system.py:2]:
```suggestion
# Copyright 2016-2017 Dirk Thomas
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright (c) 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwcli/rtwcli/command/__init__.py (hunk 4 above)
Comment (destogl) [reply rtwcli/rtwcli/rtwcli/command/__init__.py:2]:
```suggestion
# Copyright 2016-2017 Dirk Thomas
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwcli/rtwcli/command/extension_points.py (hunk 6 above)
Comment (destogl) [reply rtwcli/rtwcli/rtwcli/command/extension_points.py:2]:
```suggestion
# Copyright 2016-2017 Dirk Thomas
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwcli/rtwcli/entry_points.py (hunk 8 above)
Comment (destogl) [reply rtwcli/rtwcli/rtwcli/entry_points.py:2]:
```suggestion
# Copyright 2016-2017 Dirk Thomas
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

File: rtwcli/rtwpkg/rtwpkg/verb/__init__.py (hunk 12 above)
Comment (destogl) [reply rtwcli/rtwpkg/rtwpkg/verb/__init__.py:1]:
```suggestion
# Copyright 2017-2023 Open Source Robotics Foundation, Inc.
# Copyright 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
```

**Comment by muritane** [comment]
will be split into multiple prs

**Pull request closed without merging**